File: 0684-redundant-connection/0684-redundant-connection.py
class Solution:
    def findRedundantConnection(self, edges: List[List[int]]) -> List[int]:
        n = len(edges)
        par = [i for i in range(n+1)]
        
        def find(i):
            while par[i]!=i:
                i=par[i]
            return i
        
        for a,b in edges:
            find_a, find_b = find(a), find(b)
            if find_a != find_b:
                par[find_b] = par[find_a]
            else:
                return [a,b]
            
        return [-1, -1]
                

File: 0684-redundant-connection/0684-redundant-connection.java
class Solution {
    public int[] findRedundantConnection(int[][] edges) {
        int[] par = new int[edges.length+1];
        for(int i=0; i<par.length; ++i)
            par[i] = i;
        
        for(int[] edge: edges){
            int findA = find(par, edge[0]), findB = find(par, edge[1]);
            if(findA != findB)
                par[findB] = par[findA];
            else
                return edge;
        }
        return edges[0];
    }
    private int find(int[] par, int i){
        while(par[i] != i)
            i = par[i];
        return i;
    }
}